File: app/src/main/java/com/hsilighting/pagify/ui/components/LeavePrompt.kt
package com.hsilighting.pagify.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.hsilighting.pagify.ui.reader.LeaveIntent

/**
 * "You have unsaved marks."
 *
 * Three answers and a way out. Exit leaves them behind, Save writes them where
 * they came from, Save as writes them somewhere new — and the cross closes the
 * question without answering it, which is what a reader who pressed back by
 * accident actually wants.
 *
 * The cross rather than a Cancel button because the two are not the same idea:
 * the buttons are things to do about the marks, and closing the dialog is not one
 * of them. Sitting apart, in the corner, it reads as the dialog's own control.
 */
@Composable
fun LeavePrompt(
    intent: LeaveIntent,
    onSave: () -> Unit,
    onSaveAs: () -> Unit,
    onExit: () -> Unit,
    onClose: () -> Unit,
) {
    AlertDialog(
        onDismissRequest = onClose,
        title = {
            Row(
                Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text("Save your changes?")
                IconButton(onClick = onClose, modifier = Modifier.size(32.dp)) {
                    Icon(
                        Icons.Filled.Close,
                        contentDescription = "Close",
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }
        },
        text = {
            Text(
                when (intent) {
                    LeaveIntent.Library -> "This document has marks that have not been saved."
                    LeaveIntent.AnotherDocument ->
                        "This document has marks that have not been saved. " +
                            "Opening another will leave them behind."
                },
            )
        },
        confirmButton = { TextButton(onClick = onSave) { Text("Save") } },
        // Both remaining answers share the dismiss slot. Exit sits furthest from
        // Save so the two opposite meanings are not neighbours under a thumb.
        dismissButton = {
            Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                TextButton(onClick = onExit) { Text("Exit") }
                TextButton(onClick = onSaveAs) { Text("Save as") }
            }
        },
    )
}

File: ios/Sources/LeavePrompt.swift
import SwiftUI

/// Asked on the way out of a document with work in it.
///
/// Android's `ui/components/LeavePrompt.kt`, and the same three answers in the
/// same places. The one thing worth restating is the button order:
///
///     [ Exit ]  [ Save as ]              [ Save ]
///
/// Exit sits furthest from Save so the two opposite meanings are not neighbours
/// under a thumb. And the way out of the question is a cross rather than a
/// Cancel button, because the buttons are things to do about the marks, and
/// closing the dialog is not one of them.
struct LeavePrompt: View {
    let onSave: () -> Void
    let onSaveAs: () -> Void
    let onExit: () -> Void
    let onClose: () -> Void

    @Environment(\.colorScheme) private var scheme

    var body: some View {
        ZStack {
            // The scrim answers a tap the same way the cross does: the reader is
            // put back exactly where it was, on the same page, with every mark
            // still on it.
            Color.black.opacity(0.32)
                .ignoresSafeArea()
                .onTapGesture(perform: onClose)

            VStack(alignment: .leading, spacing: 16) {
                HStack(alignment: .firstTextBaseline) {
                    Text("Save your changes?")
                        .font(.title3.weight(.semibold))

                    Spacer(minLength: 12)

                    Button(action: onClose) {
                        Image(systemName: "xmark")
                            .font(.system(size: 15, weight: .semibold))
                            .foregroundStyle(.secondary)
                            .frame(width: 32, height: 32)
                            .contentShape(Rectangle())
                    }
                    .accessibilityLabel("Close")
                }

                Text("This document has marks that have not been saved.")
                    .font(.body)
                    .foregroundStyle(.secondary)
                    .fixedSize(horizontal: false, vertical: true)

                // The padding is the gap. Material's text buttons carry their own
                // internal padding, so a 4pt row still reads as separate actions;
                // SwiftUI's borderless buttons carry none, and "Exit Save as" ran
                // together into one phrase.
                HStack(spacing: 4) {
                    Button("Exit", action: onExit)
                        .padding(.horizontal, 10)
                    Button("Save as", action: onSaveAs)
                        .padding(.horizontal, 10)
                    Spacer(minLength: 16)
                    Button("Save", action: onSave)
                        .fontWeight(.semibold)
                        .padding(.horizontal, 10)
                }
                .buttonStyle(.borderless)
                .padding(.horizontal, -10)
            }
            .padding(20)
            .frame(maxWidth: 340)
            .background(PagifyColor.surface(scheme),
                        in: RoundedRectangle(cornerRadius: 20, style: .continuous))
            .shadow(color: .black.opacity(scheme == .dark ? 0.6 : 0.22), radius: 24, y: 8)
            .padding(24)
        }
        // The whole question is one thing to a screen reader, and the cross is
        // reachable without hunting for it.
        .accessibilityAddTraits(.isModal)
    }
}
